File: examples/chatapp/rust-be/src/gql/message.rs
use std::time::Duration;

use async_graphql::{Context, ID, Object, Result, Subscription};
use chrono::Utc;
use forgelib::Bytes;
use futures_util::{Stream, StreamExt};
use uuid::Uuid;

use crate::context::{Event, FANOUT_QUEUE, MessageJob, REAP_QUEUE, chat_topic, disappearing_secs};
use crate::db;
use crate::gql::helpers::{app, err, guarded, map_db, map_forge, me, parse_id, require_member};
use crate::gql::types::{GqlMessage, UploadTicket};

#[derive(Default)]
pub struct MessageQuery;

#[Object]
impl MessageQuery {
    async fn messages(
        &self,
        ctx: &Context<'_>,
        chat_id: ID,
        before: Option<chrono::DateTime<Utc>>,
        #[graphql(default = 50)] limit: i32,
    ) -> Result<Vec<GqlMessage>> {
        let c = app(ctx)?;
        let user = me(ctx)?;
        let cid = parse_id(&chat_id)?;
        require_member(c, cid, user.id).await?;
        let limit = limit.clamp(1, 200) as i64;
        Ok(db::list_messages(&c.pool, cid, before, limit)
            .await
            .map_err(map_db)?
            .into_iter()
            .map(GqlMessage)
            .collect())
    }
}

#[derive(Default)]
pub struct MessageMutation;

#[Object]
impl MessageMutation {
    async fn request_upload(&self, ctx: &Context<'_>, chat_id: ID) -> Result<UploadTicket> {
        let c = app(ctx)?;
        let user = me(ctx)?;
        let cid = parse_id(&chat_id)?;
        require_member(c, cid, user.id).await?;
        // fail CLOSED before minting a presign.
        let d = c
            .forge
            .ratelimit()
            .check_with(
                "upload",
                &user.id.to_string(),
                upload_limit(),
                forgelib::FailMode::Closed,
            )
            .await
            .map_err(map_forge)?;
        if !d.allowed {
            return Err(err("LIMIT", "too many upload requests; slow down"));
        }
        let max_bytes = c.max_upload_bytes().await;
        let key = format!("media/{cid}/{}", Uuid::new_v4());
        let url = c
            .forge
            .blob()
            .presign_upload(&key, Duration::from_secs(600), max_bytes)
            .await
            .map_err(map_forge)?;
        Ok(UploadTicket {
            key,
            upload_url: url,
            max_bytes: max_bytes.min(i32::MAX as u64) as i32,
        })
    }

    async fn send_message(
        &self,
        ctx: &Context<'_>,
        chat_id: ID,
        body: String,
        media_key: Option<String>,
        idempotency_key: Option<String>,
    ) -> Result<GqlMessage> {
        let c = app(ctx)?;
        let user = me(ctx)?;
        let cid = parse_id(&chat_id)?;
        require_member(c, cid, user.id).await?;

        // fail OPEN: a backend hiccup must not block messaging.
        let d = c
            .forge
            .ratelimit()
            .check_with(
                "send",
                &user.id.to_string(),
                send_limit(),
                forgelib::FailMode::Open,
            )
            .await
            .map_err(map_forge)?;
        if !d.allowed {
            return Err(err("LIMIT", "you are sending too fast; slow down"));
        }
        if body.trim().is_empty() && media_key.is_none() {
            return Err(err("INVALID", "message must have text or an attachment"));
        }

        let content_type = match &media_key {
            Some(k) => {
                if !k.starts_with(&format!("media/{cid}/")) {
                    return Err(err("INVALID", "media_key does not belong to this chat"));
                }
                c.forge
                    .blob()
                    .head(k)
                    .await
                    .map_err(map_forge)?
                    .map(|i| i.content_type)
            }
            None => None,
        };
        let expires_at = db::chat(&c.pool, cid)
            .await
            .map_err(map_db)?
            .and_then(|ch| ch.disappearing_seconds)
            .map(|s| Utc::now() + chrono::Duration::seconds(s as i64));

        let msg_id = Uuid::new_v4();

        // Client-supplied idempotency: a resend after a lost response returns the
        // original message instead of inserting a duplicate.
        if let Some(idem) = idempotency_key.as_deref().filter(|k| !k.is_empty()) {
            let key = format!("idem:send:{}:{}", user.id, idem);
            let won = c
                .forge
                .kv()
                .set(
                    &key,
                    Bytes::from(msg_id.to_string()),
                    forgelib::SetOpts::new()
                        .with_ttl(Duration::from_secs(86_400))
                        .with_mode(forgelib::SetMode::IfNotExists),
                )
                .await
                .map_err(map_forge)?;
            if !won {
                for _ in 0..5 {
                    if let Some(existing) = c.forge.kv().get(&key).await.map_err(map_forge)?
                        && let Ok(existing_id) =
                            Uuid::parse_str(&String::from_utf8_lossy(&existing))
                        && let Some(row) =
                            db::message(&c.pool, existing_id).await.map_err(map_db)?
                    {
                        return Ok(GqlMessage(row));
                    }
                    tokio::time::sleep(Duration::from_millis(50)).await;
                }
                return Err(err("INVALID", "duplicate send in progress; retry"));
            }
        }

        db::insert_message(
            &c.pool,
            msg_id,
            cid,
            user.id,
            &body,
            media_key.as_deref(),
            content_type.as_deref(),
            expires_at,
        )
        .await
        .map_err(map_db)?;

        c.publish(&chat_topic(cid), &Event::Message { message_id: msg_id })
            .await;
        enqueue_fanout(c, msg_id).await?;

        if let Some(when) = expires_at {
            let payload = serde_json::to_vec(&MessageJob { message_id: msg_id })
                .map_err(|e| err("BACKEND", e.to_string()))?;
            c.forge
                .schedule()
                .at(
                    when.into(),
                    REAP_QUEUE,
                    Bytes::from(payload),
                    forgelib::ScheduleOpts::new(),
                )
                .await
                .map_err(map_forge)?;
        }

        db::message(&c.pool, msg_id)
            .await
            .map_err(map_db)?
            .map(GqlMessage)
            .ok_or_else(|| err("BACKEND", "message vanished after insert"))
    }

    async fn set_disappearing(
        &self,
        ctx: &Context<'_>,
        chat_id: ID,
        enabled: bool,
    ) -> Result<crate::gql::types::GqlChat> {
        let c = app(ctx)?;
        let user = me(ctx)?;
        let cid = parse_id(&chat_id)?;
        require_member(c, cid, user.id).await?;
        let seconds = enabled.then(disappearing_secs);
        db::set_disappearing(&c.pool, cid, seconds)
            .await
            .map_err(map_db)?;
        // Turning off recalls not-yet-expired messages.
        if !enabled {
            db::cancel_pending_reaps(&c.pool, cid)
                .await
                .map_err(map_db)?;
        }
        db::chat(&c.pool, cid)
            .await
            .map_err(map_db)?
            .map(crate::gql::types::GqlChat)
            .ok_or_else(|| err("NOT_FOUND", "chat not found"))
    }
}

#[derive(Default)]
pub struct MessageSubscription;

#[Subscription]
impl MessageSubscription {
    async fn message_added(
        &self,
        ctx: &Context<'_>,
        chat_id: ID,
    ) -> Result<impl Stream<Item = GqlMessage> + use<>> {
        let c = app(ctx)?.clone();
        let user = me(ctx)?;
        let cid = parse_id(&chat_id)?;
        require_member(&c, cid, user.id).await?;
        let raw = c
            .forge
            .pubsub()
            .subscribe(&chat_topic(cid))
            .await
            .map_err(map_forge)?;
        let raw = guarded(c.clone(), user, raw);
        Ok(raw.filter_map(move |item| {
            let c = c.clone();
            async move {
                let bytes = item.ok()?;
                match serde_json::from_slice::<Event>(&bytes).ok()? {
                    Event::Message { message_id } => {
                        db::message(&c.pool, message_id).await.ok()?.map(GqlMessage)
                    }
                    _ => None,
                }
            }
        }))
    }
}

async fn enqueue_fanout(c: &crate::context::Ctx, message_id: Uuid) -> Result<()> {
    let payload = serde_json::to_vec(&MessageJob { message_id })
        .map_err(|e| err("BACKEND", e.to_string()))?;
    c.forge
        .queue()
        .enqueue(
            FANOUT_QUEUE,
            Bytes::from(payload),
            forgelib::EnqueueOpts::new().with_dedup_id(message_id.to_string()),
        )
        .await
        .map_err(map_forge)?;
    Ok(())
}

fn send_limit() -> forgelib::Limit {
    forgelib::Limit::per_duration(5, Duration::from_secs(10))
}

fn upload_limit() -> forgelib::Limit {
    forgelib::Limit::per_duration(30, Duration::from_secs(60))
}

File: examples/chatapp/python-be/app/gql/message.py
from __future__ import annotations

import asyncio
import json
import uuid
from collections.abc import AsyncIterator
from datetime import UTC, datetime, timedelta

import forgelib
import strawberry
from strawberry.types import Info

from .. import db
from .helpers import (
    FANOUT_QUEUE,
    INT32_MAX,
    REAP_QUEUE,
    SEND_LIMIT,
    UPLOAD_LIMIT,
    chat_topic,
    disappearing_secs,
    gqlerr,
    map_forge,
    max_upload_bytes,
    parse_id,
    publish_event,
    require_member,
    require_user,
    sub_events,
)
from .types import Chat, Message, UploadTicket, chat_from_row, message_from_row


@strawberry.type
class MessageQuery:
    @strawberry.field
    async def messages(
        self,
        info: Info,
        chat_id: strawberry.ID,
        before: datetime | None = None,
        limit: int = 50,
    ) -> list[Message]:
        u = await require_user(info)
        cid = parse_id(chat_id)
        await require_member(info, cid, u["id"])
        limit = max(1, min(200, limit))
        rows = await db.list_messages(info.context["pool"], cid, before, limit)
        return [message_from_row(r) for r in rows]


@strawberry.type
class MessageMutation:
    @strawberry.mutation(
        description="Hand the client a presigned PUT URL to upload an attachment directly to"
        " blob storage."
    )
    async def request_upload(self, info: Info, chat_id: strawberry.ID) -> UploadTicket:
        u = await require_user(info)
        cid = parse_id(chat_id)
        await require_member(info, cid, u["id"])
        forge = info.context["forge"]
        try:
            decision = await forge.rate_limit_check(
                "upload", str(u["id"]), UPLOAD_LIMIT[0], UPLOAD_LIMIT[1], fail_open=False
            )
        except forgelib.ForgeError as e:
            raise map_forge(e) from e
        if not decision.allowed:
            raise gqlerr("LIMIT", "too many upload requests; slow down")
        mb = await max_upload_bytes(info)
        key = f"media/{cid}/{uuid.uuid4()}"
        url = await forge.blob_presign_upload(key, 600, mb)
        return UploadTicket(key=key, upload_url=url, max_bytes=min(mb, INT32_MAX))

    @strawberry.mutation
    async def send_message(
        self,
        info: Info,
        chat_id: strawberry.ID,
        body: str,
        media_key: str | None = None,
        idempotency_key: str | None = None,
    ) -> Message:
        u = await require_user(info)
        pool = info.context["pool"]
        forge = info.context["forge"]
        cid = parse_id(chat_id)
        await require_member(info, cid, u["id"])

        try:
            decision = await forge.rate_limit_check(
                "send", str(u["id"]), SEND_LIMIT[0], SEND_LIMIT[1], fail_open=True
            )
        except forgelib.ForgeError as e:
            raise map_forge(e) from e
        if not decision.allowed:
            raise gqlerr("LIMIT", "you are sending too fast; slow down")
        if not body.strip() and media_key is None:
            raise gqlerr("INVALID", "message must have text or an attachment")

        content_type = None
        if media_key:
            if not media_key.startswith(f"media/{cid}/"):
                raise gqlerr("INVALID", "media_key does not belong to this chat")
            content_type = await forge.blob_content_type(media_key)

        chat_row = await db.chat(pool, cid)
        expires_at = None
        if chat_row and chat_row["disappearing_seconds"]:
            expires_at = datetime.now(UTC) + timedelta(seconds=chat_row["disappearing_seconds"])

        msg_id = uuid.uuid4()

        # Client-supplied idempotency: a resend after a lost response returns the
        # original message instead of inserting a duplicate.
        if idempotency_key:
            key = f"idem:send:{u['id']}:{idempotency_key}"
            won = await forge.kv_set(key, str(msg_id), ttl_seconds=86400, if_not_exists=True)
            if not won:
                for _ in range(5):
                    existing = await forge.kv_get(key)
                    if existing is not None:
                        row = await db.message(pool, uuid.UUID(existing))
                        if row:
                            return message_from_row(row)
                    await asyncio.sleep(0.05)
                raise gqlerr("INVALID", "duplicate send in progress; retry")

        await db.insert_message_with_receipts(
            pool, msg_id, cid, u["id"], body, media_key, content_type, expires_at
        )

        await publish_event(info, chat_topic(cid), {"type": "message", "message_id": str(msg_id)})
        # Dedup on the message id so a retried sendMessage resolver can't double-enqueue.
        await forge.queue(FANOUT_QUEUE).enqueue({"message_id": str(msg_id)}, dedup_id=str(msg_id))

        if expires_at is not None:
            await forge.schedule_at(
                expires_at.timestamp() * 1000, REAP_QUEUE, json.dumps({"message_id": str(msg_id)})
            )

        row = await db.message(pool, msg_id)
        if not row:
            raise gqlerr("BACKEND", "message vanished after insert")
        return message_from_row(row)

    @strawberry.mutation(
        description="Turn disappearing messages on/off for a chat (forge schedule)."
    )
    async def set_disappearing(self, info: Info, chat_id: strawberry.ID, enabled: bool) -> Chat:
        u = await require_user(info)
        pool = info.context["pool"]
        cid = parse_id(chat_id)
        await require_member(info, cid, u["id"])
        await db.set_disappearing(pool, cid, disappearing_secs() if enabled else None)
        if not enabled:
            # Clearing expires_at on not-yet-expired messages turns their already-scheduled
            # reap jobs into no-ops.
            await db.clear_pending_expirations(pool, cid)
        row = await db.chat(pool, cid)
        if not row:
            raise gqlerr("NOT_FOUND", "chat not found")
        return chat_from_row(row)


@strawberry.type
class MessageSubscription:
    @strawberry.subscription(description="New messages in a chat (live).")
    async def message_added(self, info: Info, chat_id: strawberry.ID) -> AsyncIterator[Message]:
        u = await require_user(info)
        pool = info.context["pool"]
        cid = parse_id(chat_id)
        await require_member(info, cid, u["id"])
        async for ev in sub_events(info, chat_topic(cid)):
            if ev.get("type") != "message":
                continue
            row = await db.message(pool, uuid.UUID(ev["message_id"]))
            if row:
                yield message_from_row(row)
